### DISCUSSION TITLE
Writting "Metadata" to new OTF Font File

### DISCUSSION THREAD

**Discussion by palmavilluan** [opentypejs/opentype.js#620] [Q&A] [answered]
i created a browser based tool whitch rasterizes the glyphs of a font input.  then the user can adjust several inputs to control the look of the new font and download the new font as OTF. to be able to recreate a specifique font i would like to write the parameter values into the font file as kind of "metadata". but im dont know how to do so.

im thankful for any help;)

**Comment by Connum** [comment, marked as answer]
You'll want to use the `meta` table, the use of which is exactly that, to add meta data. :)
According to the [name table specification](https://learn.microsoft.com/en-gb/typography/opentype/spec/meta), private tags (that is any arbitrary custom tags like in your case) MUST
> begin with an uppercase letter (0x41 to 0x5A), and must use only uppercase letters or digits

So there's no guarantee that your tag might not collide with another tag of the same name from a different font authoring software, but if you pick one that's speaking enough for your app name (as far as possible with only 4 ASCII chars), that should be relatively unlikely.

In order to add metadata to the table using opentype.js, you can do something like this:

```js
// make sure that the table exists
font.metas = font.metas || {};
font.metas.XAPP = '{"foo":"bar","baz","bat"}';
```

where `XAPP` is your custom tag name and could be anything within the constraints specified above, and the data in this case is a JSON string. To be on the safe side, you should encode all non-ASCII characters (using unicode escape sequences `\u...` in the case of JSON) in order to make sure that they can be stored and retrieved correctly, and decode them when reading the data back.
